system/lib/wasmfs/syscalls.cpp

```diff
@@ -426,4 +430,79 @@ __wasi_errno_t __wasi_fd_seek(__wasi_fd_t fd,
 
   return __WASI_ERRNO_SUCCESS;
 }
+
+long __syscall_chdir(long path) {
+  auto pathParts = splitPath((char*)path);
+
+  if (pathParts.empty()) {
+    return -ENOENT;
+  }
+
+  long err;
+  auto dir = getDir(pathParts.begin(), pathParts.end(), err);
+
+  if (!dir) {
+    return err;
+  }
+
+  wasmFS.locked().setCWD(dir);
+  return 0;
+}
+
+long __syscall_getcwd(long buf, long size) {
+  // Check if buf points to a bad address.
+  if (!buf && size > 0) {
+    return -EFAULT;
+  }
+
+  // Check if the size argument is zero and buf is not a null pointer.
+  if (buf && size == 0) {
+    return -EINVAL;
+  }
+
+  auto curr = wasmFS.locked().getCWD();
+
+  std::string buffer = "";
+
+  while (curr != wasmFS.getRootDirectory()) {
+    auto currMaybeLocked = curr->maybeLocked();
+    // TODO: better error return value.
+    // Since we are traversing up the directory tree, this checks if another
+    // thread is trying to lock a common ancestor directory.
+    // If this is the case, getcwd will not contend for access and return.
+    if (!currMaybeLocked) {
+      return -EINVAL;
+    }
+
+    // Note: getParent().lock() creates a new shared_ptr that shares ownership
+    // of the managed object. This is because parent is stored as a weak_ptr.
+    auto parentDir =
+      currMaybeLocked.value().getParent().lock()->dynCast<Directory>();
```

Comment:
This needs to handle the case that there is no parent because the CWD or one of its parents has been unlinked. Right now this will attempt to throw an exception in that case, which is not what we want. I suggest that we have `getParent` return a null `std::shared_ptr<Directory>` in this case instead.

Reply:
Could we not just check if the result of `getParent().lock()` is `nullptr`? I am assuming during `unlink` we would invalidate these pointers anyways?

Reply:
Yes, but since all users will probably call `.lock()` immediately on the result anyway, we might as well make that part of `getParent`.